### PR TITLE
Milestone: Finished PrivateEvents

### PR DESCRIPTION
**Pull request by newhorizon-tech**
This is a mini-project that allows you to

- Organize events
- Attend events
- See the list of attendees for an event
- See the events organized by the user
- See the events, a user is attending

### CODE REVIEW

File: app/models/event.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Event < ApplicationRecord
+  belongs_to :organizer, class_name: 'User'
+
+  has_many :event_attendees, foreign_key: 'event_attendee_id'
+  has_many :attendees, class_name: 'User', through: :event_attendees, source: :attendee
+
+  scope :past, -> { where('date < ?', Date.today) }
+  scope :future, -> { where('date > ?', Date.today) }
```
Comment (peterrobert) [inline app/models/event.rb]: For the future, the scope should be greater or equal to. 
`scope :upcoming, -> { where('date >= ?', Time.zone.now) }`

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED ♻️  🛑 

### Hello @newhorizon-tech  and @tnyandoro  great progress. You have shown great effort in this project. However, there is a bug in your project. And due to this, I am unable to give you a detailed review since I have no way of knowing if your project works. Kindly fix this bug and submit it for another review.

### steps.

- [x] In some projects the project works fine on your computer but not on mine. Just to be sure kindly clone your work from GitHub and if it works perfectly kindly reach me on slack @peterrobert. We shall debug together.

Screenshot from 2020-12-21 13-32-07

### if you have any question reach me on slack @peterrobert

### HAPPY CODING 💻

**Comment by newhorizon-tech** [comment]
Sorry for the inconvenience. I think the error is related to the installation of bootstrap. We used Yarn to install bootstrap. We forgot to add the instructions for installing bootstrap through Yarn in our Readme.

Please run `yarn add bootstrap jquery popper.js` and check if the project works after that

File: app/views/events/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+          </div>
+          <% if logged_in? %>
+            <% if EventAttendee.where(attend_event_id: event.id, attendee_id: current_user.id).empty? %>
+              <%= link_to 'Attend Event', attend_path(id: event), class: 'btn btn-primary' %>
+            <% else %>
+              <a href="#" class="btn btn-success">Attending</a>
+            <% end %>
+          <% else %>
+            <%= link_to 'Sign in to Attend Event', signin_path, class: 'btn btn-warning' %>
+          <% end %>
```
Comment (teekaytech) [inline app/views/events/index.html.erb]: Kindly ensure that there's no logic in views. Use helpers for view-related logic.
[Where do I put my code](http://codefol.io/posts/Where-Do-I-Put-My-Code)
[The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

File: app/views/events/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            <p>Date:
+              <%= event.date.to_date.to_formatted_s(:long_ordinal) %></p>
+          </div>
+          <% if logged_in? %>
+            <% if EventAttendee.where(attend_event_id: event.id, attendee_id: current_user.id).empty? %>
+              <a href="#" class="btn btn-warning">Did not attend</a>
+            <% else %>
+              <a href="#" class="btn btn-success">Attended</a>
+            <% end %>
+          <% end %>
```
Comment (teekaytech) [inline app/views/events/index.html.erb]: Use helpers for view-related logic.

File: app/views/events/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<div class="container">
+<h2><%= @event.title %></h2>
+<h6>Organized by <%= link_to (User.find(@event.organizer_id)).username, (User.find(@event.organizer_id))%> </h6>
+<p><%= @event.body %></p>
+
+<% if EventAttendee.where(attend_event_id: @event.id).empty? %>
```
Comment (teekaytech) [inline app/views/events/show.html.erb]: The `EventAttendee.where...` is a controller logic. Kindly add this logic to the show action of the events controller.

File: app/views/users/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+<div class="container">
+<h2><%= @user.username %>'s Profile Page</h2>
+
+<% unless @user.organize_events.empty? %>
```
Comment (teekaytech) [inline app/views/users/show.html.erb]: The `@user.organize_events` is also a controller logic. Kindly add this to the show action of your user controller.

You should get events created by the user from the controller, i.e. `@created_events = current_user.organize_events`. You can then use the @created_events in your views.

File: app/views/users/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+          <p><%= event.date %></p>
+        </div>
+      </div>
+    <% end %>
+  </div>
+</div>
+<% end %>
+
+
+<% unless EventAttendee.where(attendee_id:@user.id).empty? %>
```
Comment (teekaytech) [inline app/views/users/show.html.erb]: The `EventAttendee.where(...)` is also a controller logic. Kindly add this to the show action of your user controller.

- You can get previously attendented events by the user using event model scope, i.e. `@past_events = current_user.attended_events.past`

- Also, you can get the list of upcoming events to be attended by the user using event model scope, i.e. `@upcoming_events = current_user.attended_events.future`

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've really done well in this project.
However, some changes are required before this project can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [ ] I got your comment on using yarn to install bootstrap, jquery, and popper.js. Although this works, this is not the convention. Please find a way to use ruby gems to include these required packages in your Gemfile.
- [ ] The user's show page displays events the user has created and events that the user is attending **to be separated in past and upcoming**
- At the moment, your app displays the events the user has created and events that the user is attending but did not separate them into past and upcoming events.
- Kindly separate the events. For instance, the events that the user is attending should be separated into **past events* and **upcoming events**.
- [ ] You have some controller logic used inside your views, i.e. events and users show page. Please check my inline comments below and fix them as appropriate.
- [ ] You also have some logic embedded in your views. Check the inline comments below.



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Review by PhillipUg — APPROVED** [review]
# Status: Project Approved! 🥇 
Congratulations on implementing all requirements and getting this milestone approved. :+1:

You can merge this branch into the development branch when you're ready and go on to your next milestone.
:bulb:  Do not forget to submit your project completion form after merging this branch.

All the best team,

For any assistance, you can contact me on slack as **Phillip Musiime**

**Pull request merged**